execute/plugin_functions.go

```diff
@@ -208,6 +202,117 @@ func filterOutExecutedMessages(
 	return filtered, nil
 }
 
+// truncateObservation truncates the observation to fit within the given maxSize after encoding.
+// It removes data from the observation in the following order:
+// For each chain, remove commit reports one by one, if the encoded observation is still too large,
+// remove the entire chain.
+// Keep repeating this process until the encoded observation fits within the maxSize or there's only
+// one chain with one report left
+// Note: This function doesn't split one report into multiple parts.
+func truncateObservation(
+	observation *exectypes.Observation,
+	maxSize int,
+) {
+	encodedObs, err := observation.Encode()
+	if err != nil {
+		return
+	}
+
+	chains := maps.Keys(observation.CommitReports)
+	sort.Slice(chains, func(i, j int) bool {
+		return chains[i] < chains[j]
+	})
+
+	// If the encoded observation is too large, start filtering data.
+	for len(encodedObs) > maxSize {
+		for _, chain := range chains {
+			if len(observation.CommitReports[chain]) > 1 {
+				truncateLastCommit(chain, observation)
+			} else {
+				truncateChain(chain, observation)
+			}
+			chains = maps.Keys(observation.CommitReports)
+			break
+		}
+		// If there's only one chain with one report left, stop.
+		if len(observation.CommitReports) == 1 && len(observation.CommitReports[chains[0]]) == 1 {
```

Comment:
Hmm, so that still could overflow and be stuck forever in case of a very large single CommitReport?

Reply:
That's the subsequent PR I'm working on currently.

Reply:
Changed and erroring now when it reaches zero.